balancer/pickfirst/pickfirst.go

```diff
@@ -69,6 +77,238 @@ type pfConfig struct {
 	ShuffleAddressList bool `json:"shuffleAddressList"`
 }
 
+// subConnList stores provides functions to connect to a list of addresses using
+// the pick-first algorithm.
+type subConnList struct {
+	subConns []*scWrapper
+	// The index within the subConns list for the subConn being tried.
+	attemptingIndex int
+	b               *pickfirstBalancer
+	// The most recent failure during the initial connection attempt over the
+	// entire sunConns list.
+	lastFailure error
+	// Whether all the subConns have reported a transient failure once.
+	inTransientFailure bool
+	// Use a mutex to guard the list status as it can be changed concurrently by
+	// the idlePicker. Note that only calls to change the state to active are
+	// triggered from the picker and can happen concurrently.
+	mu     sync.RWMutex
+	status int
+}
+
+// scWrapper keeps track of the current state of the subConn.
+type scWrapper struct {
+	subConn  balancer.SubConn
+	conState connectivity.State
+	addr     resolver.Address
+}
+
+func newScWrapper(b *pickfirstBalancer, addr resolver.Address, listener func(state balancer.SubConnState)) (*scWrapper, error) {
+	scw := &scWrapper{
+		conState: connectivity.Idle,
+		addr:     addr,
+	}
+	sc, err := b.cc.NewSubConn([]resolver.Address{addr}, balancer.NewSubConnOptions{
+		StateListener: func(scs balancer.SubConnState) {
+			// Store the state and delegate.
+			scw.conState = scs.ConnectivityState
+			listener(scs)
+		},
+	})
+	if err != nil {
+		return nil, err
+	}
+	scw.subConn = sc
+	return scw, nil
+}
+
+func newSubConnList(addrs []resolver.Address, b *pickfirstBalancer) *subConnList {
+	sl := &subConnList{
+		b:      b,
+		status: subConnListPending,
+	}
+
+	for _, addr := range addrs {
+		var scw *scWrapper
+		scw, err := newScWrapper(b, addr, func(state balancer.SubConnState) {
+			sl.stateListener(scw, state)
+		})
+		if err != nil {
+			if b.logger.V(2) {
+				b.logger.Infof("Ignoring failure, could not create a subConn for address %q due to error: %v", addr, err)
+			}
+			continue
+		}
+		if b.logger.V(2) {
+			b.logger.Infof("Created a subConn for address %q", addr)
+		}
+		sl.subConns = append(sl.subConns, scw)
+	}
+	return sl
+}
+
+func (sl *subConnList) startConnectingIfNeeded() {
+	if sl == nil {
+		return
+	}
```

Comment:
Some `subConnList` methods have a `nil` receiver check while others don't. When is it possible/legal to call methods on a `nil` `subConnList`?

Reply:
The `subConnList` is nil only when the balancer hasn't received an update from the balancer yet. I've ensured that we set an empty `subConnList` in the builder and removed the `nil` receiver checks.